File: frontend/src/services/trackStatus.ts
import type { Track } from "../types/track";

const STALE_THRESHOLD_MS = 10_000;
const DROP_THRESHOLD_MS = 30_000;


export function getTrackStatus(
  lastSeen: string
): Track["status"] {

  const now = Date.now();
  const lastSeenTime = new Date(lastSeen).getTime();

  const age = now - lastSeenTime;

  if (age > DROP_THRESHOLD_MS) {
    return "DROPPED";
  }

  if (age > STALE_THRESHOLD_MS) {
    return "STALE";
  }

  return "ACTIVE";
}

File: app/services/track_status.py
from datetime import datetime, timedelta, timezone

from app.services.time_utils import ensure_utc


STALE_THRESHOLD = timedelta(seconds=10)
DROP_THRESHOLD = timedelta(seconds=30)


def get_track_status(
    last_seen: datetime,
    now: datetime | None = None
) -> str:

    if now is None:
        now = datetime.now(timezone.utc)

    last_seen = ensure_utc(last_seen)
    now = ensure_utc(now)

    age = now - last_seen

    if age >= DROP_THRESHOLD:
        return "DROPPED"

    if age >= STALE_THRESHOLD:
        return "STALE"

    return "ACTIVE"